src/goto-instrument/contracts/contracts.cpp

```diff
@@ -450,44 +489,82 @@ void code_contractst::check_apply_loop_contracts(
     converter.goto_convert(
-      monotonic_decreasing_assertion, generated_code, mode);
-    generated_code.instructions.back().source_location_nonconst().set_comment(
-      "Check decreases clause on loop iteration");
+      monotonic_decreasing_assertion, pre_loop_end_instrs, mode);
+    pre_loop_end_instrs.instructions.back()
+      .source_location_nonconst()
+      .set_comment("Check decreases clause on loop iteration");
 
     // Discard the temporary variables that store decreases clause's value
     for(size_t i = 0; i < old_decreases_vars.size(); i++)
     {
-      generated_code.add(goto_programt::make_dead(
-        old_decreases_vars[i], loop_head->source_location()));
-      generated_code.add(goto_programt::make_dead(
-        new_decreases_vars[i], loop_head->source_location()));
+      pre_loop_end_instrs.add(
+        goto_programt::make_dead(old_decreases_vars[i], loop_head_location));
+      pre_loop_end_instrs.add(
+        goto_programt::make_dead(new_decreases_vars[i], loop_head_location));
     }
   }
 
   insert_before_swap_and_advance(
     goto_function.body,
     loop_end,
-    add_pragma_disable_assigns_check(generated_code));
+    add_pragma_disable_assigns_check(pre_loop_end_instrs));
 
   // change the back edge into assume(false) or assume(guard)
   loop_end->turn_into_assume();
   loop_end->set_condition(boolean_negate(loop_end->get_condition()));
+
+  std::set<goto_programt::targett> seen_targets;
+  // Find all exit points of the loop, make temporary variables `DEAD`,
+  // and check that step case was checked for non-vacuous loops.
+  for(const auto &t : loop)
+  {
+    if(!t->is_goto())
+      continue;
+
+    auto exit_target = t->get_target();
+    if(
+      loop.contains(exit_target) ||
+      seen_targets.find(exit_target) != seen_targets.end())
+      continue;
+
+    seen_targets.insert(exit_target);
+
+    goto_programt pre_loop_exit_instrs;
+    // Assertion to check that step case was checked if we entered the loop.
+    pre_loop_exit_instrs.add(goto_programt::make_assertion(
+      or_exprt{not_exprt{entered_loop}, not_exprt{in_base_case}},
+      loop_head_location));
+    pre_loop_exit_instrs.instructions.back()
+      .source_location_nonconst()
+      .set_comment("Check that loop instrumentation was not truncated");
+    // Instructions to make all the temporaries go dead.
+    pre_loop_exit_instrs.add(
+      goto_programt::make_dead(in_base_case, loop_head_location));
+    pre_loop_exit_instrs.add(
+      goto_programt::make_dead(initial_invariant_val, loop_head_location));
+    for(const auto &v : history_var_map)
+    {
+      pre_loop_exit_instrs.add(
+        goto_programt::make_dead(to_symbol_expr(v.second), loop_head_location));
+    }
+    // Insert these instructions, preserving the loop end target.
+    insert_before_swap_and_advance(
+      goto_function.body, exit_target, pre_loop_exit_instrs);
+  }
```

Comment:
nitpick: this code could have gone into a separate function. In fact, this method has over 400 LOC and could be break down into smaller modules that would represent the major steps of our instrumentation.

Reply:
I agree. My plan is to refactor this out into `instrument_loop_contracts.cpp` in the next PR